private/bufpkg/bufcheck/bufcheckserver/internal/buflintvalidate/util.go

```diff
@@ -0,0 +1,150 @@
+	"github.com/google/cel-go/cel"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// This implements protovalidate.StandardConstraintResolver, see checkExampleValues' comment
+// for why this is needed.
+type constraintsResolverForTargetField struct {
+	protovalidate.StandardConstraintResolver
+	targetField protoreflect.FieldDescriptor
+}
+
+func (r *constraintsResolverForTargetField) ResolveFieldConstraints(desc protoreflect.FieldDescriptor) *validate.FieldConstraints {
+	if desc.FullName() != r.targetField.FullName() {
+		return nil
+	}
+	return r.StandardConstraintResolver.ResolveFieldConstraints(desc)
+}
+
+func (r *constraintsResolverForTargetField) ResolveMessageConstraints(desc protoreflect.MessageDescriptor) *validate.MessageConstraints {
+	return nil
+}
+
+func (r *constraintsResolverForTargetField) ResolveOneofConstraints(desc protoreflect.OneofDescriptor) *validate.OneofConstraints {
+	return nil
+}
+
+// This function is copied directly from protovalidate-go.
+//
+// This resolves the given extension and returns the constraints for the extension.
+func resolveExt[C proto.Message](
+	options proto.Message,
+	extType protoreflect.ExtensionType,
+) (constraints C) {
+	num := extType.TypeDescriptor().Number()
+	var msg proto.Message
+
+	proto.RangeExtensions(options, func(typ protoreflect.ExtensionType, i interface{}) bool {
+		if num != typ.TypeDescriptor().Number() {
+			return true
+		}
+		msg, _ = i.(proto.Message)
+		return false
+	})
+
+	if msg == nil {
+		return constraints
+	} else if m, ok := msg.(C); ok {
+		return m
+	}
+
+	constraints, _ = constraints.ProtoReflect().New().Interface().(C)
+	b, _ := proto.Marshal(msg)
+	_ = proto.Unmarshal(b, constraints)
+	return constraints
+}
+
+// This function is copied from protovalidate-go, with the only difference being that types
+// are based as parameters.
+//
+// reparseUnrecognized checks if there are unknown extensions on the protoreflect message.
+// If so, then it attempts to use the provided extension resolver to unmarshal the unknown
+// extension. Setting proto.UnmarshalOptions.Merge to "true" does not reset reflectMessage.Interface()
+// and allows us to unmarshal directly to reflectMessage.Interface() additively.
+func reparseUnrecognized(
+	reflectMessage protoreflect.Message,
+	extensionTypeResolver ExtensionTypeResolver,
+) error {
+	unknown := reflectMessage.GetUnknown()
+	if len(unknown) > 0 {
+		// We can call reflectMessage.SetUnknown to nil optimistically because if we fail to
+		// unmarshal, we will simply return the error.
+		reflectMessage.SetUnknown(nil)
+		options := proto.UnmarshalOptions{
+			Resolver: extensionTypeResolver,
+			Merge:    true,
+		}
+		if err := options.Unmarshal(unknown, reflectMessage.Interface()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func celTypeForStandardRuleMessageDescriptor(
+	ruleMessageDescriptor protoreflect.MessageDescriptor,
+) *cel.Type {
```

Comment:
@rodaine Just wondering if there is a better way to do this and/or if `protovalidate-go` exposes some mechanism for me to resolve the "type of rule" based on the rule message descriptor (e.g. `StringRules`). The context of this is that, we are parsing predefined rules from our files, and then checking their CEL expressions. And so we need the type for the extension itself (which is easy, we use the `celext` function you and @oliversun9 exported a while ago) and we also need the type for the rule that the extension is extending.